File: superflowsql/project.py
"""Project scaffolding - generates a full SuperFlowSQL project from templates."""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path

from jinja2 import Environment, PackageLoader, select_autoescape


@dataclass
class ProjectConfig:
    """Configuration for a new SuperFlowSQL project."""

    project_name: str = "my-superflowsql-project"
    postgres_user: str = "airflow"
    postgres_password: str = "airflow"
    postgres_db: str = "airflow"
    postgres_port: str = "5432"
    postgres_host: str = "postgres"
    airflow_user: str = "admin"
    airflow_password: str = "admin"
    airflow_email: str = "admin@example.com"
    airflow_port: str = "8080"
    airflow_fernet_key: str = "ZmDfcTF7_60GrrY167zsiPd67pEvs0aGOv2oasOM1Pg="
    pgadmin_email: str = "admin@admin.com"
    pgadmin_password: str = "admin"
    pgadmin_port: str = "5050"
    superset_user: str = "admin"
    superset_password: str = "admin"
    superset_email: str = "admin@superset.com"
    superset_port: str = "8088"
    superset_secret_key: str = "superflowsql_secret_key_change_me"

    def as_dict(self) -> dict:
        """Convert to a dict for template rendering."""
        return {k: v for k, v in self.__dict__.items()}


# Template file mapping: (template_name, output_relative_path)
_TEMPLATE_FILES: list[tuple[str, str]] = [
    ("env.j2", ".env"),
    ("docker-compose.yml.j2", "docker-compose.yml"),
    ("Dockerfile.airflow.j2", "Dockerfile.airflow"),
    ("requirements.txt.j2", "requirements.txt"),
    ("init.sql.j2", "postgres/init.sql"),
    ("pgadmin_servers.json.j2", "pgadmin_servers.json"),
    ("superset_config.py.j2", "superset/superset_config.py"),
    ("example_dag.py.j2", "dags/example_dag.py"),
]


def init_project(config: ProjectConfig, target_dir: str | Path = ".") -> Path:
    """Scaffold a new SuperFlowSQL project.

    Args:
        config: Project configuration with all service settings.
        target_dir: Parent directory in which to create the project folder.

    Returns:
        Path to the created project directory.

    Raises:
        FileExistsError: If the project directory already exists.
        OSError: If directory or file creation fails.
    """
    target_dir = Path(target_dir).resolve()
    project_dir = target_dir / config.project_name

    if project_dir.exists():
        raise FileExistsError(f"Project directory already exists: {project_dir}")

    # Create directory structure
    dirs = [
        project_dir,
        project_dir / "dags",
        project_dir / "postgres",
        project_dir / "superset",
    ]
    for d in dirs:
        d.mkdir(parents=True, exist_ok=True)

    # Render templates
    env = _get_jinja_env()
    context = config.as_dict()

    for template_name, output_path in _TEMPLATE_FILES:
        tmpl = env.get_template(template_name)
        rendered = tmpl.render(**context)
        out_file = project_dir / output_path
        out_file.parent.mkdir(parents=True, exist_ok=True)
        out_file.write_text(rendered, encoding="utf-8")

    return project_dir


def _get_jinja_env() -> Environment:
    """Create a Jinja2 environment loading from the templates package."""
    return Environment(
        loader=PackageLoader("superflowsql", "templates"),
        autoescape=select_autoescape([]),
        keep_trailing_newline=True,
        trim_blocks=True,
        lstrip_blocks=True,
    )

File: src/tui/core/project.go
package core

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"text/template"
)

// ProjectConfig holds the configuration for a new project
type ProjectConfig struct {
	ProjectName       string
	PostgresUser      string
	PostgresPassword  string
	PostgresDB        string
	PostgresPort      string
	PostgresHost      string
	AirflowUser       string
	AirflowPassword   string
	AirflowEmail      string
	AirflowPort       string
	AirflowFernetKey  string
	PgAdminEmail      string
	PgAdminPassword   string
	PgAdminPort       string
	SupersetUser      string
	SupersetPassword  string
	SupersetEmail     string
	SupersetPort      string
	SupersetSecretKey string
}

// DefaultProjectConfig returns a config with sensible defaults
func DefaultProjectConfig() ProjectConfig {
	return ProjectConfig{
		ProjectName:       "my-superflowsql-project",
		PostgresUser:      "airflow",
		PostgresPassword:  "airflow",
		PostgresDB:        "airflow",
		PostgresPort:      "5432",
		PostgresHost:      "postgres",
		AirflowUser:       "admin",
		AirflowPassword:   "admin",
		AirflowEmail:      "admin@example.com",
		AirflowPort:       "8080",
		AirflowFernetKey:  "ZmDfcTF7_60GrrY167zsiPd67pEvs0aGOv2oasOM1Pg=",
		PgAdminEmail:      "admin@admin.com",
		PgAdminPassword:   "admin",
		PgAdminPort:       "5050",
		SupersetUser:      "admin",
		SupersetPassword:  "admin",
		SupersetEmail:     "admin@superset.com",
		SupersetPort:      "8088",
		SupersetSecretKey: "superflowsql_secret_key_change_me",
	}
}

// InitProject scaffolds a new SuperFlowSQL project
func InitProject(config ProjectConfig, targetDir string) error {
	projectDir := filepath.Join(targetDir, config.ProjectName)

	// Create directory structure
	dirs := []string{
		projectDir,
		filepath.Join(projectDir, "dags"),
		filepath.Join(projectDir, "postgres"),
		filepath.Join(projectDir, "superset"),
	}

	for _, dir := range dirs {
		if err := os.MkdirAll(dir, 0755); err != nil {
			return fmt.Errorf("failed to create directory %s: %w", dir, err)
		}
	}

	// Generate files from templates
	files := map[string]string{
		filepath.Join(projectDir, ".env"):                           envTemplate,
		filepath.Join(projectDir, "docker-compose.yml"):             dockerComposeTemplate,
		filepath.Join(projectDir, "Dockerfile.airflow"):             dockerfileAirflowTemplate,
		filepath.Join(projectDir, "requirements.txt"):               requirementsTemplate,
		filepath.Join(projectDir, "postgres", "init.sql"):           initSQLTemplate,
		filepath.Join(projectDir, "pgadmin_servers.json"):           pgAdminServersTemplate,
		filepath.Join(projectDir, "superset", "superset_config.py"): supersetConfigTemplate,
		filepath.Join(projectDir, "dags", "example_dag.py"):         exampleDAGTemplate,
	}

	for filePath, tmplContent := range files {
		if err := renderTemplate(filePath, tmplContent, config); err != nil {
			return fmt.Errorf("failed to generate %s: %w", filepath.Base(filePath), err)
		}
	}

	return nil
}

func renderTemplate(filePath, tmplContent string, data interface{}) error {
	funcMap := template.FuncMap{
		"lower": strings.ToLower,
	}
	tmpl, err := template.New(filepath.Base(filePath)).Funcs(funcMap).Parse(tmplContent)
	if err != nil {
		return fmt.Errorf("template parse error: %w", err)
	}

	f, err := os.Create(filePath)
	if err != nil {
		return fmt.Errorf("file create error: %w", err)
	}
	defer f.Close()

	return tmpl.Execute(f, data)
}
